srcpkgs/python3-ipython/template

```diff
@@ -1,24 +1,35 @@
 # Template file for 'python3-ipython'
 pkgname=python3-ipython
-version=8.13.2
+version=8.14.0
 revision=1
-build_style=python3-module
-hostmakedepends="python3-setuptools"
-depends="python3-setuptools python3-jedi python3-decorator python3-pickleshare
+build_style=python3-pep517
+hostmakedepends="python3-wheel"
+depends="python3-jedi python3-decorator python3-pickleshare
  python3-traitlets python3-prompt_toolkit python3-Pygments python3-backcall
  python3-matplotlib-inline python3-pexpect python3-stack_data"
-checkdepends="python3-pytest $depends python3-matplotlib python3-Pillow
- python3-requests python3-testpath python3-numpy python3-ipython_ipykernel
- python3-jupyter_nbformat python3-Pygments"
+checkdepends="$depends python3-pytest-asyncio python3-testpath
+ python3-curio python3-jupyter_nbformat python3-matplotlib
+ python3-numpy python3-pandas python3-trio"
 short_desc="Enhanced interactive Python3 shell"
 maintainer="Andrew J. Hesford <ajh@sideband.org>"
 license="BSD-3-Clause"
 homepage="https://ipython.org/"
 changelog="https://github.com/ipython/ipython/raw/main/docs/source/whatsnew/version8.rst"
 distfiles="${PYPI_SITE}/i/ipython/ipython-${version}.tar.gz"
-checksum=7dff3fad32b97f6488e02f87b970f309d082f758d7b7fc252e3b19ee0e432dbb
+checksum=1d197b907b6ba441b692c48cf2a3a2de280dc0ac91a3405b39349a50272ca0a1
 conflicts="python-ipython<=5.8.0_2"
 
+do_check() {
+	# Running via PYTHONPATH breaks... venv works
+	# This could be moved to build-style/python3-pep517.sh
+	local testdir="${wrksrc}/.xbps-testdir"
+	rm -rf "${testdir}"
+	python3 -m venv --system-site-packages --without-pip "${testdir}"
+	local -x PATH="${testdir}/bin:${PATH}"
+	python3 -m installer dist/*.whl
+	python3 -m pytest
+}
```

Comment:
The problem isn't that this works in a `venv`, it's how you're calling `pytest`. The default `do_build` runs `pytest3` but you are running `python3 -m pytest`, which means you add `.` to the Python path by default. If you were to run
```sh
ln -Tsf /usr/bin/pytest3 "${testdir}/bin/pytest3"
"${testdir}/bin/pytest3"
```
instead of `python3 -m pytest`, the tests would again fail with an inability to find `IPython`.

This suggests that `pytest` is not actually testing against the contents of the wheel, but is instead testing the contents of the source tree. (Although, in the single test `test_ipython_embed`, it expects to find an `ipython` entrypoint and fails.) In fact, it is sufficient to do
```sh
do_check() {
    python3 -m pytest -k 'not(test_ipython_embed)'
}
```
for a successful test run.

Reply:
> The problem isn't that this works in a `venv`, it's how you're calling `pytest`. The default `do_build` runs `pytest3` but you are running `python3 -m pytest`, which means you add `.` to the Python path by default.

I see, it's really misleading that the python path changes  depending on this.

> If you were to run
> 
> ```shell
> ln -Tsf /usr/bin/pytest3 "${testdir}/bin/pytest3"
> "${testdir}/bin/pytest3"
> ```
> 
> instead of `python3 -m pytest`, the tests would again fail with an inability to find `IPython`.

I don' t think that will run pytest in the venv, since the shebang for `/usr/bin/pytest3` is `/usr/bin/python3`. It does what you claim if I do something like
```
"${testdir}/bin/python3" /usr/bin/pytest3
```
so TIL that `python3 -m pytest` is not the same as `python3 /usr/bin/pytest`.



> This suggests that `pytest` is not actually testing against the contents of the wheel, but is instead testing the contents of the source tree.

Sure and since all the test files are actually installed in the wheel, we get a conflict between the files pytest is trying to load (from `.`) and the modules that are installed in `$testdir`.

So an easy way to make `.` into the python path would be
```
make_check_pre='env PYTHONPATH=.'
```

With this in place and the standard `do_check()`, all tests pass (even `test_ipython_embed`). What do you think?

Reply:
I think that's the best option, because it requires a single line without overriding the default functions.